ghc-lib-gen/src/Ghclibgen.hs

```diff
@@ -1037,70 +1093,21 @@ applyPatchCmmParseNoImplicitPrelude _ = do
-                "Cabal-3.8.1.0"
-              , "Cabal-syntax-3.8.1.0"
-              ]] ++
-        [ file | ghcSeries ghcFlavor >= GHC_9_8
-          , file <- [
-                "unix-2.8.5.0"
-              , "directory-1.3.8.2"
-              , "process-1.6.18.0"
-              , "filepath-1.4.100.4"
-              , "Win32-2.13.4.0"
-              , "time-1.12.2"
-              , "semaphore-compat-1.0.0"
-        ] ] ++
-        case parse (\cfg -> cfg .:? "extra-deps" .!= []) config of
-          Success ls -> ls :: [Y.Value]
-          Error msg -> error msg
-  -- Build hadrian (and any artifacts we generate via hadrian e.g.
-  -- Parser.hs) as quickly as possible.
-  let opts = HMS.insert "$everything" "-O0 -j" $
-        case parse (\cfg -> cfg .:? "ghc-options" .!= HMS.empty) config of
-          Success os -> os :: HMS.HashMap T.Text Y.Value
-          Error msg -> error msg
-  let config' =
-        HMS.insert "extra-deps" (toJSON deps)
-          (HMS.insert "ghc-options" (toJSON opts)
-#if !MIN_VERSION_aeson(2, 0, 2)
-                                     config
-#else
-                                     (toHashMap config)
-#endif
-          )
-    -- [Note: Hack "ghc-X.X.X does not compile with ghc XXX"]
-    -- ------------------------------------------------------
-    -- See for example
-    -- https://gitlab.haskell.org/ghc/ghc/-/issues/21633 &
-    -- https://gitlab.haskell.org/ghc/ghc/-/issues/21634.
-    --
-    -- The idea is to replace the resolver with whatever is prevailing
-    -- (or ghc-9.6.4 if that's not possible).
-    -- 9.6 since 2024/02/26 d9d69e127a735cfc3dbe8b1f7dc96a06fe654c3e
-      resolverDefault = "lts-22.12" -- ghc-9.6.4
-      -- The resolver has to curate packages so resolvers of the form
-      -- ghc-x.y.z won't do.
-      resolver' = case fromMaybe resolverDefault resolver of
-        r | "ghc-" `isPrefixOf` r -> resolverDefault
-        r -> r
-      -- This is still an issue with 9.6.1 (resolver in
-      -- hadrian/stack.yaml is a 9.0.2 resolver i.e. not recent enough
-      -- to build GHC so causes a configure error).
-      config'' = if ghcSeries ghcFlavor < GHC_9_4
-                     then config'
-                     else
-                           -- Ignore 'hadrian.cabal' constraints.
-                           HMS.insert "allow-newer" (toJSON True) $
-                           HMS.update (\_ -> Just (toJSON resolver')) "resolver" config'
-  Y.encodeFile hadrianStackYaml config''
+applyPatchHadrianCabalProject :: GhcFlavor -> IO ()
+applyPatchHadrianCabalProject _ = do
+    whenM (doesPathExist "hadrian/cabal.project.freeze") $
+      removePathForcibly "hadrian/cabal.project.freeze"
+    cabalProjectContents <- lines' <$> readFile' "hadrian/cabal.project"
+    writeFile "hadrian/cabal.project" $
+      unlines ( cabalProjectContents ++
+                [
+                  "flags: -selftest"
+                , "if impl(ghc < 9.0)"
+                , "  allow-newer: all"
```

Comment:
that's a good question. i didn't drill into that. i just found empirically that for the 8.8.1 builds, if i don't relax the solver we get build errors. in this case in fact a compile error: 
```
src/Hadrian/Haskell/Cabal/Parse.hs:75:58: error:
    • Couldn't match type ‘[Char]’ with ‘C.ShortText’
      Expected type: C.ShortText
        Actual type: String
    • In the second argument of ‘($)’, namely ‘C.synopsis pd’
      In the third argument of ‘PackageData’, namely
        ‘(C.fromShortText $ C.synopsis pd)’
      In the second argument of ‘($)’, namely
        ‘PackageData
           name
           version
           (C.fromShortText $ C.synopsis pd)
           (C.fromShortText $ C.description pd)
           depPkgs
           gpd’
   |
75 |     return $ PackageData name version (C.fromShortText $ C.synopsis pd) (C.fromShortText $ C.description pd) depPkgs gpd
   |                                                          ^^^^^^^^^^^^^
```
similarly, line 1110 where i remove the lines that start with "index-state" which are  in effect "freezing" dependencies to the state of hackage on some particular date. these are removed so as to relax those constraints too though not doing that leads to failures to find build plans (which is more "normal") rather than compile errors.

Reply:
could it relate to this DA modification?
```
#ifdef WITH_BAZEL
    return $ PackageData name version (C.fromShortText $ C.synopsis pd) (C.fromShortText $ C.description pd) depPkgs gpd
#else
    return $ PackageData name version (C.synopsis pd) (C.description pd) depPkgs gpd
#endif
```

Reply:
i am going to push a new revision relating to this in this PR.

Reply:
Great to hear that we only needed `QuickCheck` and `Cabal` - I was worried it'd be a massive list. Thanks :)